File: typescript/difference-of-squares/difference-of-squares.ts
export default class Squares {
    private readonly target: number
    squareOfSum: number
    sumOfSquares: number
    difference: number

    constructor(target: number) {
        this.target = target
        this.squareOfSum = this._squareOfSum()
        this.sumOfSquares = this._sumOfSquares()
        this.difference = Math.abs(this.sumOfSquares - this.squareOfSum)
    }

    private _squareOfSum(): number {
        let res = 0
        for (let i = 0; i <= this.target; i++) {
            res += i
        }
        return res * res
    }

    private _sumOfSquares(): number {
        let res = 0
        for (let i = 0; i <= this.target; i++) {
            res += i * i
        }
        return res

    }
}

File: go/difference-of-squares/difference_of_squares.go
package diffsquares

import "math"

// SquareOfSum takes a int as input and return the square of sum.
// formula: 1+2+3+...+n=n*(n+1)/2
func SquareOfSum(n int) (result int) {
	sum := n * (n + 1) / 2
	return sum * sum
}

// SumOfSquares takes a int as input and return the sum of square.
// formula: 1*1+2*2+3*3+...+n*n=n*(n+1)*(2n+1)/6
func SumOfSquares(n int) (result int) {
	return n * (n + 1) * (2*n + 1) / 6
}

// Difference function would return the absolute difference between SquareOfSum and SumOfSquares
func Difference(target int) int {
	return abs(SumOfSquares(target) - SquareOfSum(target))

}

func abs(n int) int {
	return int(math.Abs(float64(n)))
}
